torch/csrc/autograd/functions/init.cpp

```diff
@@ -62,27 +63,83 @@ struct NoCtor {
 };
 
 template<typename C, typename T>
-static void addClass(PyObject* module, PyTypeObject& type, const char* name)
+static void addClass(PyObject* module, PyTypeObject& type, const char* name,
+  PyGetSetDef* function_properties=NULL, PyMethodDef* function_methods=NULL)
 {
-  createForwardFunctionPyTypeObject<T>(type, name);
+  createForwardFunctionPyTypeObject<T>(type, name, function_properties, function_methods);
   Py_INCREF(&type);
   PyModule_AddObject(module, name, (PyObject*)&type);
   registerCppFunction(typeid(C), &type);
 }
 
+template<typename T, typename M, typename P, M P::*ptr, typename V, PyObject* (*Convert)(V)>
+PyObject* getTupleAttr(PyObject* obj, void* _unused)
+{
+  THPCppFunction* self = (THPCppFunction*)obj;
+  auto& arr = std::static_pointer_cast<T>(self->cdata).get()->*ptr;
```

Comment:
Wouldn't it be simpler to first `.get()` the pointer, and then use a cast on that (no need to shared_ptr casts then)?